### PR TITLE
Raise an  error for unsupported ctx managers

### PR DESCRIPTION
**Pull request by nakulcr7**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108250

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

Differential Revision: [D48836688](https://our.internmc.facebook.com/intern/diff/D48836688)

### CODE REVIEW

**Comment by nakulcr7** [comment]
@nakulcr7 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48834550).

File: torch/_dynamo/variables/ctx_manager.py [hunk 1]
```diff
@@ -78,30 +78,36 @@ def enter(self, tx):
...
-            **options,
-        ).call_function(tx, [], {})
+        try:
+            return variables.UserMethodVariable(
+                self.cm_obj.__enter__.__func__,
+                variables.UserDefinedObjectVariable(self.cm_obj, **options),
+                **options,
+            ).call_function(tx, [], {})
+        except Exception:
+            unimplemented(f"context manager `{self.cm_obj}`")
```
Comment (anijain2305) [inline torch/_dynamo/variables/ctx_manager.py:88]: Can we customize the error message better here?

Ideal way would be to catch the raised exception here, print the original error message, and then add a new error message that this is an unsupported context manager enter or exit function.

Comment (nakulcr7) [reply torch/_dynamo/variables/ctx_manager.py]: @anijain2305 I've added a `log.exception` to capture the original error message. Lmk if that's too verbose

**Comment by nakulcr7** [comment]
@nakulcr7 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D48836688).

File: torch/_dynamo/variables/ctx_manager.py [hunk 2]
```diff
@@ -78,30 +81,38 @@ def enter(self, tx):
...
-            variables.UserDefinedObjectVariable(self.cm_obj, **options),
-            **options,
-        ).call_function(tx, [], {})
+        try:
+            return variables.UserMethodVariable(
+                self.cm_obj.__enter__.__func__,
+                variables.UserDefinedObjectVariable(self.cm_obj, **options),
+                **options,
+            ).call_function(tx, [], {})
+        except Exception as e:
```
Comment (anijain2305) [inline torch/_dynamo/variables/ctx_manager.py:90]: So, we can have various type of exceptions here. We want to be more specific. We can do something like this

```
        except Unsupported as e:
            raise unimplemented(f"Unsupported context manager {self.cm_obj}'s __enter__ function") from e
```

1) Ensure we catch `Unsupported` only exceptions
2) Chaining exceptions is a better pythonic style.

Comment (anijain2305) [reply torch/_dynamo/variables/ctx_manager.py:90]: Apart from that this PR LGTM.

**Comment by nakulcr7** [comment]
Commit chain got messed up. Will open up a new PR

**Pull request closed without merging**